File: Fortran/ex4/fmincg.f90
!double precision function fmincg(length, nn_params, input_layer_size, hidden_layer_size, num_labels, inputdata, y, lambda)
   
!	subroutine fmincg(M, N, X, y, nn_params)
	subroutine fmincg(M,N,M1,N1,M2,N2, nn_params,X, y,lambda, length)
       
	use ieee_arithmetic
	implicit none
	 
	integer::M, N, M1,N1,M2,N2
	double precision,dimension(M,N)::X		 
	double precision,dimension(M,1)::y		  	 
    
    ! Copyright (C) 2001 and 2002 by Carl Edward Rasmussen. Date 2002-02-13
    ! (C) Copyright 1999, 2000 & 2001, Carl Edward Rasmussen
    ! 
    ! Permission is granted for anyone to copy, use, or modify these
    ! programs and accompanying documents for purposes of research or
    ! education, provided this copyright notice is retained, and note is
    ! made of any changes that have been made.
    ! 
    ! These programs and documents are distributed without any warranty,
    ! express or implied.  As the programs were written for research
    ! purposes only, they have not been tested to the degree that would be
    ! advisable in any important application.  All use of these programs is
    ! entirely at the user's own risk.
    !
    ! [ml-class] Changes Made:
    ! 1) Function name and argument specifications
    ! 2) Output display
    !
    ! [John Shahbazian http://fortrandev.wordpress.com 5/31/2014] Changes Made:
    ! 1) Ported to Fortran
    ! 2) Changed the cost function call to internal.  Replace the
    !    'costandgradient' function to whatever you would like.  It returns
    !    the cost as the result, and the gradient is returned through the first 
    !    argument as intent(inout) (e.g. 'gradient#').  
    ! 3) Changed the variable names to be readable.
    !    f1 = cost1
    !    df1 = grad1
    !    s = search_direction
    !    d1 = slope1
    !    z1 = point1
    !    X0 = backup_params
    !    f0 = cost_backup
    !    df0 = gradient_backup    
    
!    integer, intent(in) :: length,input_layer_size,hidden_layer_size ,num_labels
!    integer :: length
!      double precision, allocatable, intent(in) :: inputdata(:,:), y(:,:)
!      double precision, allocatable  :: inputdata(:,:)
!    double precision, intent(in) :: lambda
	double precision:: lambda
    double precision:: nn_param(M1*N1+M2*N2)
	double precision:: nn_params(M1*N1+M2*N2, 1)

    
    double precision :: RHO = 0.01                            ! a bunch of constants for line searches
    double precision :: SIG = 0.5       ! RHO and SIG are the constants in the Wolfe-Powell conditions
    double precision :: INT = 0.1    ! don't reevaluate within 0.1 of the limit of the current bracket
    double precision :: EXT = 3.0                    ! extrapolate maximum 3 times the current bracket
    integer :: MAXEVALS = 20                         ! max 20 function evaluations per line search
    double precision :: RATIO = 100                                      ! maximum allowed slope ratio

    double precision :: mintemp, minstuff, MM, A, B
    double precision :: fX = 0.0

    integer :: success
    integer :: costFunctionCount  ,length                                       ! zero the run length counter
    integer :: ls_failed = 0                                    ! no previous line search has failed

    double precision, allocatable :: backup_params(:,:)
    double precision :: cost1, cost2, cost3, cost_backup
    double precision, allocatable :: grad1(:,:), grad2(:,:), gradient3(:,:), gradient_backup(:,:), tmp(:,:)
    double precision :: limit
    double precision :: point1, point2, point3, ptemp
    double precision, allocatable :: search_direction(:,:), stemp(:,:)
    double precision :: slope1, slope2, slope3, slope_vector(1,1)
    double precision :: sqrtnumber
    double precision :: sd_calc_1(1,1), sd_calc_2(1,1), sd_calc_3(1,1), sd_calc_4
    double precision, allocatable :: sd_calc_5(:,:)


    allocate(grad1(1:size(nn_params),1:1))
    allocate(grad2(1:size(nn_params),1:1))
    allocate(gradient3(1:size(nn_params),1:1))
    allocate(gradient_backup(1:size(nn_params),1:1))
    allocate(tmp(1:size(nn_params),1:1))
    allocate(search_direction(1:size(nn_params),1:1))
    allocate(stemp(1:size(nn_params),1:1))
    allocate(backup_params(1:size(nn_params),1:1))
	allocate(sd_calc_5(size(search_direction,1),size(search_direction,2)))
	

!	 length = 50
	 costFunctionCount=0  !Very important
	 

!    cost1 = costandgradient(grad1,nn_params,input_layer_size,hidden_layer_size,num_labels, inputdata, y, lambda)

	  call nnCostFunction(M,N,M1,N1,M2,N2, nn_params, X, y, lambda, cost1, grad1)
!      if(length < 0)then
!        costFunctionCount = costFunctionCount + 1    !(length<0)                                          ! count epochs?!
!      endif
	
    search_direction = -grad1                               ! search direction is steepest
    slope_vector = matmul(-transpose(search_direction),search_direction)    ! this is the slope
    slope1 = slope_vector(1,1)                                  !convert to scalar
    point1  = 1.0/(1.0-slope1)                                  ! initial step is red/(|s|+1)
	
    do while(costFunctionCount < abs(length)) 
                                            
!        costFunctionCount = costFunctionCount + (length>0)                                      ! count iterations?!
	 if(length > 0)then
	    costFunctionCount = costFunctionCount+1
	 endif
        backup_params = nn_params
        cost_backup = cost1
        gradient_backup = grad1                             ! make a copy of current values
        stemp = point1  * search_direction
        nn_params = nn_params + stemp                           ! begin line search
	  
        grad2 = grad1
	    
!        cost2 = costandgradient(grad2,nn_params,input_layer_size,hidden_layer_size,num_labels, inputdata, y, lambda)
	 call nnCostFunction(M,N,M1,N1,M2,N2, nn_params, X, y, lambda, cost2, grad2)
	 !Pass too many parameters to subroutine will slow down the computing

        if(length < 0)then
          costFunctionCount = costFunctionCount + 1    !(length<0)                                          ! count epochs?!
        endif
!        costFunctionCount = costFunctionCount + (length<0)                                      ! count epochs?!

        slope_vector = matmul(transpose(grad2),search_direction)              ! this is the slope
        slope2 = slope_vector(1,1)                              !convert to scalar

        cost3 = cost1
        slope3 = slope1
        point3 = -point1                                        ! initialize point 3 equal to point 1
        if(length > 0)then
            MM = MAXEVALS
        else
            MM = min(MAXEVALS, -length-costFunctionCount)
        end if
        success = 0
        limit = -1                                              ! initialize quantities
	  
        do
            do while (( (cost2 > (cost1 + (point1 * RHO * slope1))) .or. (slope2 > (-SIG * slope1)) ) .and. (MM > 0) )
                limit = point1                                  ! tighten the bracket
                if(cost2 > cost1)then
                    point2 = point3 - (0.5 * slope3 * point3 * point3)/(slope3 * point3 + cost2 - cost3)                 ! quadratic fit
                else
                    A = 6*(cost2 - cost3)/point3 + 3*(slope2 + slope3)                                 ! cubic fit
                    B = 3*(cost3 - cost2) - point3 * (slope3 + 2*slope2)
                    point2 = (sqrt(B*B - A*slope2*point3*point3) - B)/A
                end if
                if(ieee_is_nan(point2) .or. (.not. ieee_is_finite(point2)))then
                    point2 = point3 / 2                         ! if we had a numerical problem then bisect
                end if
                point2 = MAX( MIN(point2, (INT * point3)), ((1.0 - INT) * point3))  ! don't accept too close to limits
                point1  = point1 + point2                       ! update the step
                stemp = point2 * search_direction
                nn_params = nn_params + stemp

!                cost2 = costandgradient(grad2,nn_params,input_layer_size,hidden_layer_size,num_labels, inputdata, y, lambda)
		   call nnCostFunction(M,N,M1,N1,M2,N2, nn_params, X, y, lambda, cost2, grad2)
		   
                MM = MM - 1
!                costFunctionCount = costFunctionCount + (length<0)                              ! count epochs?!
                  if(length < 0)then
        			costFunctionCount = costFunctionCount + 1    !(length<0)                                          ! count epochs?!
      		endif

                slope_vector = matmul(transpose(grad2),search_direction)
                slope2 = slope_vector(1,1)                      !convert to scalar
                point3 = point3 - point2                        ! point3 is now relative to the location of point2
            end do
	 
!	      write(*,*)'cost2=', cost2, 'cost1=', cost1
		
            if ((cost2 > (cost1 + (point1*RHO*slope1)) ) .or. (slope2 > (-SIG * slope1) ))then
                exit                                            ! this is a failure
            elseif (slope2 > (SIG * slope1))then
                success = 1
                exit                                            ! success
            elseif (MM == 0)then
                exit                                            ! failure
            end if
            A = 6*(cost2 - cost3)/point3 + 3*(slope2 + slope3)  ! make cubic extrapolation
            B = 3*(cost3 - cost2) - point3*(slope3 + 2*slope2)
            sqrtnumber = (B*B) - A*slope2*point3*point3
            if((.not. ieee_is_normal(sqrtnumber)) .or. sqrtnumber < 0 )then
                if (limit < -0.5) then                          ! if we have no upper limit
                    point2 = point1  * (EXT - 1)                ! the extrapolate the maximum amount
                else
                    point2 = (limit - point1)/2                 ! otherwise bisect
                end if
            else
                point2 = (-slope2 * point3 * point3)/(B + sqrt(sqrtnumber))
                if ((limit > -0.5) .and. ((point2 + point1) > limit))then          ! extraplation beyond max?
                    point2 = (limit - point1)/2                 ! bisect
                elseif ((limit < -0.5) .and. ((point2 + point1) > (point1 * EXT)))then       ! extrapolation beyond limit
                    point2 = point1 * (EXT - 1.0)               ! set to extrapolation limit
                elseif (point2 < (-point3 * INT))then
                    point2 = -point3 * INT
                elseif ((limit > -0.5) .and. (point2 < (limit - point1)*(1.0 - INT)))then   ! too close to limit?
                    point2 = (limit - point1 ) * (1.0 - INT)
                end if
            end if

            cost3 = cost2
            slope3 = slope2
            point3 = -point2               
            point1  = point1  + point2

            stemp = point2 * search_direction
            nn_params = nn_params + stemp                       ! update current estimates

 !           cost2 = costandgradient(grad2,nn_params,input_layer_size,hidden_layer_size,num_labels, inputdata, y, lambda)
		call nnCostFunction(M,N,M1,N1,M2,N2, nn_params, X, y, lambda, cost2, grad2)
	 
            MM = MM - 1
!            costFunctionCount = costFunctionCount + (length<0)                                  ! count epochs?!
		if(length < 0)then
        		costFunctionCount = costFunctionCount + 1    !(length<0)                                          ! count epochs?!
      	endif
            slope_vector = matmul(transpose(grad2),search_direction)
            slope2 = slope_vector(1,1)                          !convert to scalar
        end do                                                  ! end of line search

		
        if (success == 1)then                                   ! if line search succeeded
            cost1 = cost2
            fX = cost1
			write(*,*)'sucess Cost =', cost1

            ! this calculation was split up to make it easier to work with
            sd_calc_1 = matmul(transpose(grad2),grad2)
            sd_calc_2 = matmul(transpose(grad1),grad2)
            sd_calc_3 = matmul(transpose(grad1),grad1)
            sd_calc_4 = (sd_calc_1(1,1) - sd_calc_2(1,1)) / sd_calc_3(1,1)
            sd_calc_5 = sd_calc_4 * search_direction
            search_direction = sd_calc_5 - grad2

            tmp = grad1
            grad1 = grad2
            grad2 = tmp                                     ! swap derivatives
            slope_vector = matmul(transpose(grad1),search_direction)
            slope2 = slope_vector(1,1)                          !convert to scalar
            if(slope2 > 0)then                                  ! new slope must be negative
                search_direction = -grad1                   ! otherwise use steepest direction
                slope_vector = matmul(-transpose(search_direction),search_direction)
                slope2 = slope_vector(1,1)                      !convert to scalar
            end if
            mintemp = slope1/(slope2 - 2.2251D-308)  !2.2551D-308 is min value double precision float
            minstuff = min(RATIO, mintemp)
            point1  = point1  * minstuff                         ! slope ratio but max RATIO
            slope1 = slope2
            ls_failed = 0                                        ! this line search did not fail
        else
            nn_params = backup_params
            cost1 = cost_backup
            grad1 = gradient_backup                         ! restore point from before failed line search
            if (ls_failed == 1 .or. costFunctionCount > abs(length))then        ! line search failed twice in a row
                exit                                            ! or we ran out of time, so we give up
            end if
            tmp = grad1
            grad1 = grad2
            grad2 = tmp                                    ! swap derivatives
            search_direction = -grad1                      ! try steepest
            slope_vector = matmul(-transpose(search_direction),search_direction)
            slope1 = slope_vector(1,1)                         ! convert to scalar
            point1  = 1.0 / (1.0 - slope1)
            ls_failed = 1                                      ! this line search failed
        end if

    end do


!    fmincg = fX  
!end function fmincg
end subroutine fmincg

File: C/ex2/fmincg.c
/*
Copyright (C) 2001 and 2002 by Carl Edward Rasmussen. Date 2002-02-13
%
%
% (C) Copyright 1999, 2000 & 2001, Carl Edward Rasmussen
% 
% Permission is granted for anyone to copy, use, or modify these
% programs and accompanying documents for purposes of research or
% education, provided this copyright notice is retained, and note is
% made of any changes that have been made.
% 
% These programs and documents are distributed without any warranty,
% express or implied.  As the programs were written for research
% purposes only, they have not been tested to the degree that would be
% advisable in any important application.  All use of these programs is
% entirely at the user's own risk.

[ Sagar GV, 2013, sagar.writeme@gmail.com ] Changes Made:
- Ported to C
*/
//#include "fmincg.h"

#include<math.h>
#include<stdio.h>
#include<float.h>

//#define COST_FUNC_DATATYPE double
//#define COST_FUNC_DATATYPE_MIN (FLT_MIN*100)

#define RHO 0.01f
#define SIG 0.5f
#define INT 0.1f
#define EXT 3.0f
#define MAX 20
#define RATIO 100.0f

int fmincg(double (*lrCostFunction)(int M, int N, double theta[N], double X[M][N], double y[M], double grad[N], double lambda), int M, int N, double X[M][N], double y[M], double theta[N], double lambda, int maxCostCalls)
{
	int i,success = 0,costFuncCount=0,lineSearchFuncCount=0;
	double ls_failed,Cost1,d1,z1,f0,Cost2,d2,f3,d3,z3,limit,z2,A,B,C;
	double grad1[N],s[N],x0[N],df0[N],grad2[N],tmp[N];
	
	int j;

	ls_failed = 0;

	if(costFuncCount >= maxCostCalls) return 1; else costFuncCount++;
	Cost1 = lrCostFunction(M, N, theta, X, y, grad1, lambda);
	

	for(i=0;i<N;i++)
		s[i] = -grad1[i];
	
	d1 = 0;
	for(i=0;i<N;i++)
		d1 += -s[i]*s[i];

	z1 = 1.0f / (1 - d1);
	
	while(1)
	{
		for(i=0;i<N;i++){
			x0[i] = theta[i];
			df0[i] = grad1[i];
		}
		f0 = Cost1;
		
		for(i=0;i<N;i++){
			theta[i] = theta[i] + (z1)*s[i];
		}
		if(costFuncCount >= maxCostCalls) return 1; else costFuncCount++;
		Cost2 = lrCostFunction(M, N, theta, X, y, grad2, lambda);


		d2 = 0;
		for(i=0;i<N;i++)
			d2 += grad2[i]*s[i];
					
		f3 = Cost1;
		d3 = d1;
		z3 = -z1;
		
		success = 0; 
		limit = -1;
		lineSearchFuncCount = 0;
		// begin line search
		while(1){
			while((( (Cost2) > ((Cost1) + RHO*(z1)*(d1))) || ( (d2) > -SIG*(d1) )) && lineSearchFuncCount < MAX){
				limit = z1;
				if( (Cost2) > (Cost1) ){
					z2 = z3 - (0.5f*(d3)*(z3)*(z3))/((d3)*(z3)+(Cost2)-(f3));
				}
				else{
					A = 6*((Cost2)-(f3))/(z3)+3*((d2)+(d3));
					B = 3*((f3)-(Cost2))-(z3)*((d3)+2*(d2));
					z2 = (sqrt(B*B-A*(d2)*(z3)*(z3))-B)/A;
				}
				if(isnan(z2) || isinf(z2)){
					z2 = (z3) * 0.5f;
				}
				
				A = ((z2 < INT*(z3)) ? z2 : INT*(z3));
				B = (1-INT)*(z3);
				z2 = A > B ? A : B;
				z1 = z1 + z2;
				
				for(i=0;i<N;i++)
					theta[i] += (z2)*s[i];

				if(costFuncCount >= maxCostCalls) return 1; else costFuncCount++;
				lineSearchFuncCount++;
				Cost2 = lrCostFunction(M, N, theta, X, y, grad2, lambda);

				d2 = 0;
				for(i=0;i<N;i++)
					d2 += grad2[i] * s[i];
				
				z3 = z3 - z2;
			}

			if( (Cost2 > Cost1 + (z1)*RHO*(d1)) || ((d2) > -SIG*(d1)) ){
				break; //failure
			}else if( d2 > SIG*(d1) ){
				success = 1; break; 
			}else if(lineSearchFuncCount >= MAX){
				break;
			}
			
			A = 6*(Cost2-f3)/z3+3*(d2+d3);
			B = 3*(f3-Cost2)-z3*(d3+2*d2);
			z2 = -d2*z3*z3/(B+sqrt(B*B-A*d2*z3*z3));
			if(!(B*B-A*d2*z3*z3 >= 0) || isnan(z2) || isinf(z2) || z2 < 0){
				if(limit < -0.5f){
					z2 = z1 * (EXT-1);
				}else{
					z2 = (limit-z1)/2;
				}
			}else if((limit > -0.5) && (z2+z1 > limit)){
				z2 = (limit-z1)/2; 
			}else if((limit < -0.5) && (z2+z1 > z1*EXT)){
				z2 = z1*(EXT-1.0);
			}else if(z2 < -z3*INT){
				z2 = -z3*INT;
			}else if((limit > -0.5) && (z2 < (limit-z1)*(1.0-INT))){
				z2 = (limit-z1)*(1.0-INT);
			}
			
			f3 = Cost2; d3 = d2; z3 = -z2;
			z1 = z1 + z2;
			for(i=0;i<N;i++)
				theta[i] += z2*s[i];
			
			if(costFuncCount >= maxCostCalls) return 1; else costFuncCount++;
			lineSearchFuncCount++;
			Cost2 = lrCostFunction(M, N, theta, X, y, grad2, lambda);
			d2 = 0;
			
			for(i=0;i<N;i++)
				d2 += grad2[i]*s[i];
			
		}
		// line search ended
		if(success){
			Cost1 = Cost2;
//			printf("Cost: %e\n", Cost1);
			
			A = 0;
			B = 0;
			C = 0;
			for(i=0;i<N;i++){
				A += grad1[i]*grad1[i];
				B += grad2[i]*grad2[i];
				C += grad1[i]*grad2[i];
			}
			for(i=0;i<N;i++){
				s[i] = ((B-C)/A)*s[i] - grad2[i];
			}
			for(i=0;i<N;i++){
				tmp[i] = grad1[i]; grad1[i] = grad2[i]; grad2[i] = tmp[i];
			}
			d2 = 0;
			for(i=0;i<N;i++)
				d2 += grad1[i] * s[i];
			
			if(d2 > 0){
				for(i=0;i<N;i++)
					s[i] = -grad1[i];
				
				d2 = 0;
				for(i=0;i<N;i++)
					d2 += -s[i]*s[i];				
			}
			
//			A = d1/(d2-COST_FUNC_DATATYPE_MIN);
			A = d1/(d2- 0.);
			z1 = z1 * ((RATIO < A) ? RATIO : A);
			d1 = d2;
			ls_failed = 0;
		}else{
			Cost1 = f0;
			for(i=0;i<N;i++){
				theta[i] = x0[i];
				grad1[i] = df0[i];
			}
			
			if(ls_failed){
				break;
			}
			
			for(i=0;i<N;i++)
				tmp[i] = grad1[i]; grad1[i] = grad2[i]; grad2[i] = tmp[i];
			
			for(i=0;i<N;i++)
				s[i] = -grad1[i]; 
			
			d1 = 0;
			for(i=0;i<N;i++)
				d1 += -s[i]*s[i];
			
			z1 = 1/(1-d1);
			ls_failed = 1;
		}
	}
	return 2;
}
